flatten.go

```diff
@@ -0,0 +1,167 @@
+package authzed
+
+import (
+	"fmt"
+
+	api "github.com/authzed/authzed-go/arrakisapi/api"
+)
+
+// FlattenExpand reduces an ExpandResponse into the slice of Users present in
+// the expansion.
+//
+// Notably, this removes the context of which relations caused users to be
+// included in the expansion, but often you only need to know which users are
+// present.
+func FlattenExpand(resp *api.ExpandResponse) []*api.User { return flatten(resp.TreeNode) }
+
+func flatten(node *api.RelationTupleTreeNode) []*api.User {
+	switch typed := node.NodeType.(type) {
+	case *api.RelationTupleTreeNode_IntermediateNode:
+		switch typed.IntermediateNode.Operation {
+		case api.SetOperationUserset_UNION:
+			return flattenUnion(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_INTERSECTION:
+			return flattenIntersection(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_EXCLUSION:
+			return flattenExclusion(typed.IntermediateNode.ChildNodes)
+		}
+	case *api.RelationTupleTreeNode_LeafNode:
+		users := newUserSet()
+		for _, usr := range typed.LeafNode.Users {
+			users.add(usr)
+		}
+		return users.toSlice()
+	}
+	return nil
+}
+
+func flattenUnion(children []*api.RelationTupleTreeNode) []*api.User {
+	users := newUserSet()
+	for _, child := range children {
+		users.add(flatten(child)...)
+	}
+	return users.toSlice()
+}
+
+func flattenIntersection(children []*api.RelationTupleTreeNode) []*api.User {
+	firstChildChildren := flatten(children[0])
+
+	if len(children) == 1 {
+		return firstChildChildren
+	}
+
+	inOthers := newUserSet()
+	inOthers.add(flattenIntersection(children[1:])...)
+
+	maxChildren := len(firstChildChildren)
+	if len(inOthers) < maxChildren {
+		maxChildren = len(inOthers)
+	}
+
+	toReturn := make([]*api.User, 0, maxChildren)
+	for _, child := range firstChildChildren {
```

Comment:
Perhaps add `intersect` and `subtract` to the user set itself?

Reply:
Those are both operations on slices right now to avoid allocations. I think we're best with what we've got.